slow_conv3d: Avoid dispatch in parallel region

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66165
* #65759
* #65758
* #65757
* __->__ #65737
* #65726

See gh-56794

Avoid dispatch inside of parallel_for by:
- Replacing Tensor slicing with TensorAccessor
- Copy bias into output only once, outside of the parallel region
- Replaces `addmm_` and `baddbmm_` with direct calls to gemm.

Differential Revision: [D31257874](https://our.internmc.facebook.com/intern/diff/D31257874)

peterbell10 committed:
slow_conv3d: Avoid dispatch in parallel region

See gh-56794

Avoid dispatch inside of parallel_for by:
- Replacing Tensor slicing with TensorAccessor
- Copy bias into output only once, outside of the parallel region
- Replaces `addmm_` and `baddbmm_` with direct calls to gemm.

[ghstack-poisoned]

peterbell10 committed:
Update on "slow_conv3d: Avoid dispatch in parallel region"


See gh-56794

Avoid dispatch inside of parallel_for by:
- Replacing Tensor slicing with TensorAccessor
- Copy bias into output only once, outside of the parallel region
- Replaces `addmm_` and `baddbmm_` with direct calls to gemm.

[ghstack-poisoned]

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31257874).

peterbell10 committed:
Update on "slow_conv3d: Avoid dispatch in parallel region"


See gh-56794

Avoid dispatch inside of parallel_for by:
- Replacing Tensor slicing with TensorAccessor
- Copy bias into output only once, outside of the parallel region
- Replaces `addmm_` and `baddbmm_` with direct calls to gemm.

Differential Revision: [D31257874](https://our.internmc.facebook.com/intern/diff/D31257874)

[ghstack-poisoned]

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31257874).

peterbell10 committed:
Update on "slow_conv3d: Avoid dispatch in parallel region"


See gh-56794

Avoid dispatch inside of parallel_for by:
- Replacing Tensor slicing with TensorAccessor
- Copy bias into output only once, outside of the parallel region
- Replaces `addmm_` and `baddbmm_` with direct calls to gemm.

Differential Revision: [D31257874](https://our.internmc.facebook.com/intern/diff/D31257874)

[ghstack-poisoned]

peterbell10 committed:
Update on "slow_conv3d: Avoid dispatch in parallel region"


See gh-56794

Avoid dispatch inside of parallel_for by:
- Replacing Tensor slicing with TensorAccessor
- Copy bias into output only once, outside of the parallel region
- Replaces `addmm_` and `baddbmm_` with direct calls to gemm.

Differential Revision: [D31257874](https://our.internmc.facebook.com/intern/diff/D31257874)

[ghstack-poisoned]

peterbell10 committed:
Update on "slow_conv3d: Avoid dispatch in parallel region"


See gh-56794

Avoid dispatch inside of parallel_for by:
- Replacing Tensor slicing with TensorAccessor
- Copy bias into output only once, outside of the parallel region
- Replaces `addmm_` and `baddbmm_` with direct calls to gemm.

Differential Revision: [D31257874](https://our.internmc.facebook.com/intern/diff/D31257874)

[ghstack-poisoned]

peterbell10 committed:
Update on "slow_conv3d: Avoid dispatch in parallel region"


See gh-56794

Avoid dispatch inside of parallel_for by:
- Replacing Tensor slicing with TensorAccessor
- Copy bias into output only once, outside of the parallel region
- Replaces `addmm_` and `baddbmm_` with direct calls to gemm.

Differential Revision: [D31257874](https://our.internmc.facebook.com/intern/diff/D31257874)

[ghstack-poisoned]

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31257874).

ngimel commented on aten/src/ATen/native/mkl/LinearAlgebra.cpp:
```diff
@@ -100,6 +133,63 @@ static inline void gemm_batched(const CBLAS_TRANSPOSE trans_A, const CBLAS_TRANS
+  TORCH_INTERNAL_ASSERT(false, "Unknown TransposeType");
+}
+
+void mkl_gemm_batched(
```
it seems these functions are the same as `gemm_batched` above, modulo type of trans_A/trans_B, why do we need 2 copies?

peterbell10 replied:
The biggest difference is `ColMajor` vs. `RowMajor`. In a follow up I can remove `_baddbmm_mkl_` entirely though. Just have the normal version call `cpublas::gemm_batched_with_stride`.

ngimel commented on aten/src/ATen/native/CPUBlas.cpp:
```diff
@@ -251,6 +255,154 @@ void gemm(
+    scalar_t beta,
+    scalar_t **c, int64_t ldc) {
+  if (batch_size == 1) {
+    return gemm(transa, transb, m, n, k, alpha, a[0], lda, b[0], ldb, beta, c[0], ldc);
```
this is nice! I was wondering if calling gemm_batched regardles of `groups` is going to be ok, but with this special case it is

ngimel reviewed (commented):
Thanks! Adding batched gemm to at::native::cpublas is great, I have a small question about code duplication
